docs/plugin/messaging.md

```diff
@@ -0,0 +1,886 @@
+
+```c++
+#define SYSLOG(CATEGORY, PARAMETERS)
+    do {
+        static_assert(std::is_base_of<WPEFramework::Logging::BaseLoggingType<CATEGORY>, CATEGORY>::value, "SYSLOG() only for Logging controls");
+        if (CATEGORY::IsEnabled() == true) {
+            CATEGORY __data__ PARAMETERS;
+            WPEFramework::Core::Messaging::MessageInfo __info__(
+                CATEGORY::Metadata(),
+                WPEFramework::Core::Time::Now().Ticks()
+            );
+            WPEFramework::Core::Messaging::IStore::Logging __log__(__info__);
+            WPEFramework::Messaging::TextMessage __message__(__data__.Data());
+            WPEFramework::Messaging::MessageUnit::Instance().Push(__log__, &__message__);
+        }
+    } while(false)
+```
+
+In addition, you might be wondering why we have some `Messaging` components such as `Metadata` in `Core`, and why everything is not simply inside
+`Source/messaging`. From an architectural point of view, there is a reason for this, which revolves around the need for reporting capabilities to be accessible not only in the plugins and other Thunder components, but also within `Core`, where, for instance, we would like to measure how long it takes to lock and then unlock. To accomplish this, we need some of the messaging features to be accessible in `Core`, because otherwise we would get, as you may have guessed, circular dependencies.
+
+
+# Warning Reporting
+
+Last but not least, it is about time to describe warning reporting. It is a crucial aspect of software systems that aims to alert developers and users about potential issues or anomalies within the system's operation. Unlike logging and tracing, which primarily focus on capturing and storing detailed information for diagnostic purposes, warning reporting specifically targets situations where certain conditions might lead to unexpected behavior or errors. While logging records events and activities to provide a comprehensive record of system activity, and tracing follows the flow of execution across different components or services, warning reporting is designed to raise flags about specific conditions that could lead to failures, performance degradation, or security vulnerabilities.
+
+By emphasizing the significance of these conditions, warning reporting enables proactive identification and resolution of potential problems, enhancing system reliability and user experience. It acts as an early detection mechanism, signaling the need for attention and potential action before the situation escalates into a critical failure or incident. Overall, warning reporting complements logging and tracing by focusing on identifying and communicating conditions that require immediate attention, facilitating effective troubleshooting and maintenance of software systems.
+
+```c++
+REPORT_OUTOFBOUNDS_WARNING(WarningReporting::SinkStillHasReference, _referenceCount);
```

Comment:
Is this line meant to be here since we go through all the macros just below it?

Reply:
Good catch, it's indeed no longer necessary, as I expanded upon this example in the next section, so I'll remove this line.